File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ParcelSelectObjects.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.UUID;

/**
 * ParcelSelectObjects
 * viewer -> sim
 * reliable
 *
 * <p>Template: {@code ParcelSelectObjects Low 202 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
public class ParcelSelectObjects extends SLMessage {
    public AgentData AgentData_Field;
    public ParcelData ParcelData_Field;
    public ArrayList<ReturnIDs> ReturnIDs_Fields = new ArrayList<>();

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public UUID SessionID; // LLUUID
    }

    /** Block ParcelData, Single. */
    public static class ParcelData {
        public int LocalID; // S32
        public int ReturnType; // U32
    }

    /** Block ReturnIDs, Variable. */
    public static class ReturnIDs {
        public UUID ReturnID; // LLUUID
    }

    public ParcelSelectObjects() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.ParcelData_Field = new ParcelData();
    }

    @Override
    public int CalcPayloadSize() {
        return (this.ReturnIDs_Fields.size() * 16) + 45;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleParcelSelectObjects(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 202 (ParcelSelectObjects).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0xCA);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packInt(byteBuffer, this.ParcelData_Field.LocalID);
        packInt(byteBuffer, this.ParcelData_Field.ReturnType);
        byteBuffer.put((byte) this.ReturnIDs_Fields.size());
        Iterator<?> it = this.ReturnIDs_Fields.iterator();
        while (it.hasNext()) {
            packUUID(byteBuffer, ((ReturnIDs) it.next()).ReturnID);
        }
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.ParcelData_Field.LocalID = unpackInt(byteBuffer);
        this.ParcelData_Field.ReturnType = unpackInt(byteBuffer);
        int i = byteBuffer.get() & 0xFF;
        for (int i2 = 0; i2 < i; i2++) {
            ReturnIDs returnIDs = new ReturnIDs();
            returnIDs.ReturnID = unpackUUID(byteBuffer);
            this.ReturnIDs_Fields.add(returnIDs);
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ParcelSelectObjects.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.UUID

/**
 * ParcelSelectObjects
 * viewer -> sim
 * reliable
 *
 * <p>Template: {@code ParcelSelectObjects Low 202 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
open class ParcelSelectObjects : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var ParcelData_Field: ParcelData = ParcelData()
    @JvmField val ReturnIDs_Fields = ArrayList<ReturnIDs>()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block ParcelData, Single. */
    open class ParcelData {
        @JvmField var LocalID: Int = 0
        @JvmField var ReturnType: Int = 0
    }

    /** Block ReturnIDs, Variable. */
    open class ReturnIDs {
        @JvmField var ReturnID: UUID? = null
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return (ReturnIDs_Fields.size * 16) + 45
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleParcelSelectObjects(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 202 (ParcelSelectObjects).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0xCA).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packInt(byteBuffer, ParcelData_Field.LocalID)
        packInt(byteBuffer, ParcelData_Field.ReturnType)
        byteBuffer.put((ReturnIDs_Fields.size.toByte()))
        for (entry in ReturnIDs_Fields) {
            packUUID(byteBuffer, entry.ReturnID)
        }
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        ParcelData_Field.LocalID = unpackInt(byteBuffer)
        ParcelData_Field.ReturnType = unpackInt(byteBuffer)
        val i = (byteBuffer.get().toInt() and 0xFF)
        repeat(i) {
            val returnIDs = ReturnIDs()
            returnIDs.ReturnID = unpackUUID(byteBuffer)
            ReturnIDs_Fields.add(returnIDs)
        }
    }
}
